Sources/SpeziLocalLLM/LLMLlama+Generation.swift

```diff
@@ -0,0 +1,167 @@
+// SPDX-FileCopyrightText: 2023 Stanford University
+//
+// SPDX-License-Identifier: MIT
+//
+
+import Foundation
+import llama
+
+
+/// Extension of ``LLMLlama`` handling the text generation.
+extension LLMLlama {
+    /// Typealias for the llama.cpp `llama_token`.
+    typealias LLMLlamaToken = llama_token
+    
+    
+    /// Based on the input prompt, generate the output with llama.cpp
+    ///
+    /// - Parameters:
+    ///   - prompt: The input `String` prompt.
+    ///   - continuation: A Swift `AsyncThrowingStream` that streams the generated output.
+    func _generate( // swiftlint:disable:this identifier_name function_body_length
+        prompt: String,
+        continuation: AsyncThrowingStream<String, Error>.Continuation
+    ) {
+        self.state = .generating
+        
+        let tokens = tokenize(text: prompt)
+        let nKVReq = UInt32(tokens.count) + UInt32((self.parameters.nLength - Int(tokens.count)))
+        
+        let context = llama_new_context_with_model(model, self.contextParameters.wrapped)
+        guard context != nil else {
+            print("Failed to initialize context")
+            continuation.finish(throwing: LLMError.generationError)
+            return
+        }
+        defer {
+            llama_free(context)
+        }
+
+        let nCtx = llama_n_ctx(context)
+        
+        print("\n_length = \(self.parameters.nLength), n_ctx = \(nCtx), n_batch = \(self.contextParameters.nBatch), n_kv_req = \(nKVReq)\n")
+
+        if nKVReq > nCtx {
+            print("error: n_kv_req (%d) > n_ctx, the required KV cache size is not big enough\n", nKVReq)
+            continuation.finish(throwing: LLMError.generationError)
+            return
+        }
+        
+        var buffer: [CChar] = []
+        for id: LLMLlamaToken in tokens {
+            print(token_to_piece(token: id, buffer: &buffer) ?? "", terminator: "")
+        }
+        
+        print("\n")
+        
+        var batch = llama_batch_init(Int32(tokens.count), 0, 1)
+        defer {
+            llama_batch_free(batch)
+        }
+        
+        // evaluate the initial prompt
+        batch.n_tokens = Int32(tokens.count)
+        
+        for (tokenIndex, token) in tokens.enumerated() {
+            batch.token[tokenIndex] = token
+            batch.pos[tokenIndex] = Int32(tokenIndex)
+            batch.n_seq_id[tokenIndex] = 1
+            batch.seq_id[tokenIndex]?[0] = 0
+            batch.logits[tokenIndex] = 0
+        }
+        
+        // llama_decode will output logits only for the last token of the prompt
+        batch.logits[Int(batch.n_tokens) - 1] = 1
+        
+        if llama_decode(context, batch) != 0 {
+            print("llama_decode() failed")
+            continuation.finish(throwing: LLMError.generationError)
+            return
+        }
+        
+        var nCur = batch.n_tokens
+        var nDecode = 0
+        
+        var streamBuffer: [CChar] = .init()
+
+        let startTime = ggml_time_us()
```

Comment:
I am wondering if just using Swift date where would also be sufficient. Is there an advantage to use the C++/ObjC functions here?